proposals/0000-mutable-fields.rst

```diff
@@ -0,0 +1,249 @@
+.. proposal-number::
+
+.. trac-ticket::
+
+.. implemented::
+
+Mutable Constructor Fields
+==========================
+
+We propose an extension to add support for mutable constructor fields.
+
+Motivation
+----------
+
+This proposal improves both the runtime and memory overhead for code
+that currently uses ``IORef`` or ``STRef``.
+
+The best representation we currently have available for a type that
+contains an ``IORef`` is something like::
+
+  data MutPair a b =
+    MutPair {-# UNPACK #-} !(IORef a)
+            {-# UNPACK #-} !(IORef b)
+
+and the runtime representation of this is 3 heap objects: the outer
+``MutPair`` with 2 pointer fields (3 words total), and 2 ``MutVar#``
+objects (2 words each).
+
+Under this proposal, we would be able to declare a ``MutPair`` that is
+represented by a single 3-word heap object containing 2 directly
+mutable pointer fields.
+
+
+Proposed Change
+---------------
+
+We propose the addition of a new primitive type constructor::
+
+  Ref# :: * -> * -> TYPE 'Ref
+
+(``Ref`` is a new ``RuntimeRep``) and two new primitives::
+
+  readRef#  :: Ref# s a -> State# s -> (# State# s, a #)
+  writeRef# :: Ref# s a -> a -> State# s -> (# State# s, () #)
+
+A ``Ref#`` represents a mutable field of a constructor.  Although ``Ref#``
+appears to be a normal first-class primitive type, its *runtime
+representation* will be ``(# Any, Int# #)``, that is, an unboxed pair
+of the object that contains the mutable field and the offset of the
+mutable field.
+
+
+Defining a type with mutable fields
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+First, let's make a couple of handy type synonyms::
+
+  type IOField a = Ref# RealWorld a
+  type STField s a = Ref# s a
+
+The only way to create a ``Ref#`` is by declaring a constructor that has a
+``Ref#`` field.  For example (using the ``IOField`` synonym)::
+
+  data MutPair a b = MutPair (IOField a) (IOField b)
+
+A ``MutPair`` is a constructor with two mutable fields.  But ``MutPair``
+is not a constructor in the ordinary sense; both allocation and
+pattern-matching behave in a special way.
+
+The ``MutPair`` constructor has the following type::
+
+  MutPair :: a -> b -> IO (MutPair a b)
```

Comment:
It seems very odd to me to change the type of the constructor like this; this feels like a very ugly exception where the moment you add a mutable field the behavior of your constructor changes in a very fundamental way. It almost feels as though there should be some syntactic indicator to mark these special constructors (e.g. `data MutPair = mutable MutPair (IOField a) (IOField b)`, which could also be used in the GADT case).

I tried playing around with a few alternative approaches to provide the "constructor action" via a typeclass (such that we could just hide the constructor itself), but it sadly seems to get quite complicated. My initial thought was that you could identify a constructor with its promoted kind (e.g. `'MutPair`), but the presence of the field arguments in its kind make this rather awkward. To make this fly we'd need to have a better way to represent constructor identity at the type level.

Another approach would be to build this on the existing `Generics` infrastructure, although this also seemed quite complicated and not necessarily safe.

Regardless, while changing the type of the constructor itself is awkward, it may be the least of the evils here.